### ISSUE TITLE
pm2 startup failed

### ISSUE THREAD

**Issue by liyunsheng** [Unitech/pm2#1416]
run any command, even `pm2 -v`, always print `[PM2] Spawning PM2 daemon` and then quit.
in `$HOME/.pm2/logs`show the reason:

```
PM2 v0.14.3: God Daemon: ../deps/uv/src/unix/linux-core.c:746: read_times: Assertion `num == numcpus' failed.
```

my machine is a VPS, cat /proc/stat  show this
![image](https://cloud.githubusercontent.com/assets/9405946/8517265/c2bbdd34-23f1-11e5-9658-404a98ab672f.png)

**Is there any way you can solve this problem?** thanks

Environment info
![image](https://cloud.githubusercontent.com/assets/9405946/8517329/7a16f4aa-23f2-11e5-8e4d-e1c45bbb214e.png)
cat /proc/cpuinfo  show 24 CPU
node  v0.12.5
pm2 v0.14.3

**Comment by jshkurti** [comment]
https://github.com/Unitech/PM2/issues/1067 duplicate

**Comment by jshkurti** [comment]
I think you should post an issue on node.js or io.js github.
This isn't related to PM2 in my opinion.

**Comment by soyuka** [comment]
+1 @jshkurti [source](https://github.com/joyent/node/blob/c146855f06f7c6ed41b53329a5d814e89b6eaa15/deps/uv/src/unix/linux-core.c#L746)

**Comment by Unitech** [comment]
Can you re run any PM2 commands like that: `DEBUG="*" pm2 my-command` and paste output here
Thanks

**Comment by itsatony** [comment]
i have the same problem on 1 of our servers. we virtualize using lxc. i suspect it having to do something with the container, have not found the setting yet though. i tested (on all machines) with 
node 0.12.x 4.2.1 5.0.0

the failing machine logs this with DEBUG=* :

itsatony@dev-toni:~/code/deployDemo$ DEBUG=* pm2 --no-daemon
  pm2:constants PM2_ROOT_PATH: /home/itsatony/.pm2 +0ms
  pm2:monit Getting interaction info +62ms
  pm2:interface:daemon [PING INTERACTOR] Trying to connect to Interactor daemon +2ms
  axon:sock connect attempt null:/home/itsatony/.pm2/interactor.sock +0ms
pm2 launched in no-daemon mode (you can add DEBUG="*" env variable to get more messages)
  pm2:satan [PING PM2] Trying to connect to server +6ms
  axon:sock connect attempt null:/home/itsatony/.pm2/rpc.sock +5ms
  axon:sock error ENOENT +3ms
  axon:sock ignored ENOENT +0ms
  axon:sock error ENOENT +0ms
  axon:sock ignored ENOENT +0ms
  axon:sock attempting reconnect +100ms
  axon:sock closing +1ms
  axon:sock closing 0 connections +0ms
  pm2:satan Daemon not launched +105ms
  axon:sock connect attempt null:/home/itsatony/.pm2/rpc.sock +1ms
  pm2:satan [PING PM2] Trying to connect to server +1ms
  axon:sock connect attempt null:/home/itsatony/.pm2/rpc.sock +1ms
  axon:sock attempting reconnect +0ms
  axon:sock closing +0ms
  axon:sock closing 0 connections +0ms
  pm2:interface:daemon Interactor Daemon not launched +1ms
  axon:sock connect attempt null:/home/itsatony/.pm2/interactor.sock +1ms
  axon:sock error ENOENT +0ms
  axon:sock ignored ENOENT +0ms
  axon:sock error ENOENT +0ms
  axon:sock ignored ENOENT +0ms
  axon:sock error ENOENT +1ms
  axon:sock ignored ENOENT +0ms
  axon:sock attempting reconnect +100ms
  axon:sock closing +0ms
  axon:sock closing 0 connections +0ms
  pm2:satan Daemon not launched +103ms
  axon:sock connect attempt null:/home/itsatony/.pm2/rpc.sock +1ms
  pm2:satan Launching in no daemon mode +0ms
node: ../deps/uv/src/unix/linux-core.c:746: read_times: Assertion `num == numcpus' failed.
Aborted (core dumped)

**Comment by itsatony** [comment]
my (heroic) sysop has problably identified the problem (changing the setting made pm2 start). 
we are using lxc virtualization and that part of the node-code does not work with a limiting of the cpu cores.

the relevant container setting is:
    cgroup setting: cpuset.cpus
explained here: https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Resource_Management_Guide/sec-cpuset.html

I hope that helps others as well.. if you are running on virtualized servers, ask the hoster to adjust the values (they probably wont, as that would mean unlimiting your cpu-count in the vm).

P.S. : Similar problems might occur when running in limited docker containers.

**Comment by robbmanes** [comment]
If you're hitting this issue in either a VM or a container, this is probably fixed by a recent `libuv` patch into `v1.48`:  https://github.com/libuv/libuv/commit/434eb4b0ac29e853e85562119fda3fa681642f17

`libuv` in `nodejs` will no longer compare the `/proc/stat` and `/proc/cpuinfo` in the same way, now properly accounting for when cpusets are applied to VM's and/or if VM's are quickly starting up and running `pm2` or other `nodejs` tasks when they come online.  This is done in the form of a hotplug-compatibility patch since the Linux kernel supports activating/deactivating CPU's more-or-less at will.

It got pulled into `nodejs` just two weeks ago and my own preliminary testing with `podman` and `kata-containers` to simulate the VM seems to show it resolves the issue:  https://github.com/nodejs/node/commit/36dcd399c05fe4f25c759e5c54aac9c10d2b98f8

The bleeding-edge `nodejs` at this moment contains `libuv` with this patch, so if you're seeing this problem I'd encourage you to update your `nodejs` version to absolute latest and try again.  At the time of writing this, it's not even included in a tag yet, so I don't imagine any distributed version has it:
```bash
$ git remote -v
origin  https://github.com/nodejs/node.git (fetch)
origin  https://github.com/nodejs/node.git (push)

$ git tag --contains=36dcd399c05fe4f25c759e5c54aac9c10d2b98f8 | wc -l
0
```
